Firestore/core/src/firebase/firestore/core/transaction_runner.mm

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/core/transaction_runner.h"
+#include "Firestore/core/src/firebase/firestore/remote/exponential_backoff.h"
+#include "absl/algorithm/container.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+namespace {
+
+using remote::RemoteStore;
+using util::AsyncQueue;
+using util::TimerId;
+using util::Status;
+
+/** Maximum number of times a transaction can be retried before failing. */
+constexpr int kRetryCount = 5;
+
+bool IsRetryableTransactionError(const util::Status& error) {
+  // In transactions, the backend will fail outdated reads with
+  // FAILED_PRECONDITION and non-matching document versions with ABORTED. These
+  // errors should be retried.
+  Error code = error.code();
+  return code == Error::Aborted || code == Error::FailedPrecondition ||
+         !remote::Datastore::IsPermanentError(error);
+}
+}  // namespace
+
+TransactionRunner::TransactionRunner(const std::shared_ptr<AsyncQueue>& queue,
+                                     RemoteStore* remote_store,
+                                     TransactionUpdateCallback update_callback,
+                                     TransactionResultCallback result_callback)
+    : queue_{queue},
+      remote_store_{remote_store},
+      update_callback_{update_callback},
+      result_callback_{result_callback},
+      backoff_{queue_, TimerId::RetryTransaction},
+      retries_left_{kRetryCount} {
```

Comment:
I'm surprised you didn't get a build error with this change, but you should initialize these in the same order they're declared.